### PR TITLE
Add NequIP architecture

### PR DESCRIPTION
**Pull request by Satvik24511**
## Description

This draft PR proposes the implementation of the NequIP model.

The PR focuses only on the main model architecture. It does not add the
`TorchModel` training wrapper, energy-force loss, loader integration, etc.

The implementation accepts torch-converted DeepChem `GraphData` or
`BatchGraphData` and returns structure-level energies. Conservative forces can
optionally be computed as the negative gradient of total energy with respect
to atomic positions.

### Motivation

DeepChem already contains useful equivariant building blocks, but its existing
`Fiber` and `SE3GraphConv` abstractions do not directly represent the complete
original NequIP architecture:

- `Fiber` records angular degree `l` but not O(3) parity;
- `SE3GraphConv` uses different radial, path-mixing, aggregation, and residual
  semantics;
- NequIP requires parity-aware gated nonlinearities and a
  species-conditioned self connection;
- NequIP keeps compatible tensor-product paths separate until a subsequent
  atom-wise linear transformation.

This implementation reuses DeepChem primitives where their mathematics match
while adding only the NequIP-specific bookkeeping and operations required by
the paper.

No new external dependency is introduced. In particular, the implementation
does not require e3nn, DGL, or torch-geometric.

## Architecture

The backbone follows this data flow:

```text
GraphData / BatchGraphData
        |
atomic numbers + positions + edge_index + graph_index
        |
atomic-number embedding into even scalar (0e) features
        |
differentiable edge geometry reconstructed from positions
        |
Bessel radial basis × polynomial cutoff
        |
real spherical harmonics
        |
NequIPInteractionBlock × num_interaction_blocks
        |
even-scalar atom-wise readout
        |
per-atom energies
        |
graph_index sum
        |
structure energies
        |
optional -dE/dR
        |
conservative forces
```

### O(3) feature representation

Node features are stored internally as:

```python
(l, parity) -> tensor
```

where:

- `l` is the angular degree;
- parity is `+1` for even irreps and `-1` for odd irreps;
- each tensor has shape
  `(num_atoms, multiplicity, 2*l + 1)`.

This small internal representation adds the parity information absent from
DeepChem's existing `Fiber` class without introducing a new general-purpose
irreps framework.

### Differentiable geometry

The graph topology is treated as fixed, but all geometric quantities used by
the energy calculation are reconstructed inside `forward()`:

```python
center = edge_index[0]
neighbor = edge_index[1]
r_ij = positions[neighbor] - positions[center]
```

Distances, directions, radial features, and spherical harmonics therefore
remain connected to `node_pos_features` through autograd.

Stored `edge_features` and `edge_distances` are intentionally ignored by the
backbone. This prevents NumPy-derived static geometry from breaking force
gradients.

### Radial representation

The edge-distance representation follows NequIP equation 6:

- trainable Bessel frequencies initialized to `n*pi`;
- prefactor `2 / cutoff`;
- polynomial cutoff with `p=6`;
- numerically safe sinc evaluation of the `r -> 0` limit;
- per-interaction radial MLPs;
- normalized SiLU hidden activations;
- no output activation.

The radial MLP produces independent per-edge weights for every allowed
tensor-product path.

### Tensor-product convolution

For each input irrep, spherical-harmonic filter degree, and requested output
irrep, the implementation enforces:

```text
abs(l_in - l_filter) <= l_out <= l_in + l_filter
p_out = p_in * (-1)^l_filter
```

The convolution follows the `uvu` multiplicity semantics used by the
paper-era NequIP implementation:

- input multiplicity is preserved within a path;
- edge spherical harmonics have multiplicity one;
- one radial weight is generated per input channel and path;
- paths producing the same output irrep remain separate;
- paths are concatenated only before the post-convolution atom-wise linear
  transformation.

Messages read the neighbor feature from `edge_index[1]`, aggregate into the
central atom at `edge_index[0]`, use SUM aggregation, and are normalized by:

```text
sqrt(avg_num_neighbors)
```

`avg_num_neighbors` is an explicit model parameter. The model does not replace
it with node degree or a batch-derived statistic.

### DeepChem spherical-harmonic and CG reuse

The implementation reuses:

- `get_spherical_from_cartesian`;
- `precompute_sh`;
- `basis_transformation_Q_J`.

DeepChem's spherical harmonics are integral-normalized and its `Q_J` basis is
unit-normalized. The tensor-product contraction applies:

```text
sqrt(4*pi*(2*l_out + 1))
```

to convert this basis to the component normalization used by the paper-era
NequIP/e3nn tensor product.

The allowed paths and normalization were checked for all angular couplings
through `l_max=2`.

### Interaction block

Each `NequIPInteractionBlock` applies the paper ordering:

```text
residual = species_self_connection(x)
x = pre_self_interaction(x)
x = tensor_product_convolution(x)
x = post_self_interaction(x)  # part of the convolution module
x = x + residual
x = equivariant_gate(x)
```

The species-conditioned self connection is represented by one channel matrix
per atomic species and compatible `(l, parity)` irrep. Each matrix acts
identically over the irrep's `m` components, which is equivalent to contracting
the input features with one-hot species attributes.

The gate uses the paper-era nonlinearities:

- even scalar features: normalized SiLU;
- odd scalar features: normalized tanh;
- higher-order features: one normalized even scalar gate per channel.

### Energy readout and forces

Only even `l=0` features enter the output block. Two atom-wise equivariant
linear layers produce one scalar atomic energy per atom, with no activation
between them.

Atomic energies are summed by `BatchGraphData.graph_index`, producing:

```text
(batch_size, 1)
```

When requested, forces are calculated without a direct-force head:

```python
forces = -torch.autograd.grad(total_energy.sum(), positions)
```

`create_graph=True` is supported so a later `TorchModel` integration can train
against force targets.

## Existing DeepChem components reused

This PR reuses the following existing components rather than duplicating them:

- `GraphData`
- `BatchGraphData`
- `BatchGraphData.graph_index`
- `Fiber`
- `SE3SelfInteraction`
- `get_spherical_from_cartesian`
- `precompute_sh`
- `basis_transformation_Q_J`
- standard PyTorch embedding, tensor operations, and autograd

`SE3SelfInteraction` is applied separately to even and odd feature groups
because `Fiber` itself does not track parity.

## New NequIP-specific components

The new code is limited to the operations not already represented faithfully
by existing DeepChem abstractions:

- lightweight `(l, parity)` bookkeeping;
- NequIP Bessel basis and polynomial cutoff;
- radial MLP with paper-era variance scaling;
- explicit parity-aware tensor-product path construction;
- NequIP tensor-product aggregation and normalization;
- species-conditioned self connection;
- parity-aware equivariant gate;
- interaction-block composition;
- scalar energy readout and optional conservative-force calculation.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be 0.32.0**)
  - [x] Run `mypy -p deepchem` and check no 
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
    - The new module contains API documentation but no executable doctest
      examples.
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
  - Public class and method documentation is included in `nequip.py`.
- [x] I have added tests that prove my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by ARY2260** [comment]
@JoseAntonioSiguenza Do you plan to add support for "parity" handling in the existing SE3SelfInteraction and Fiber classes? Also can you please comment on the approach here.
